nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,446 @@
+					return *addr.IpAddress, nil
+				}
+				return *addr.Ipv6Address, nil
+			}
+			v4, v6 := getHTTPSIPFromNodeEntity(node.Entities)
+			if v4 != "" {
+				return v4, nil
+			}
+			return v6, nil
+		}
+	}
+	return "", fmt.Errorf("Failed to get host ip from cluster config")
+}
+
+func getHTTPSIPFromNodeEntity(entities []nsxModel.NodeEntityInfo) (string, string) {
+	// returns the ipv4 and ipv6 address of a node's https port 443
+	for _, entity := range entities {
+		if *entity.Port == 443 {
+			return *entity.IpAddress, *entity.Ipv6Address
+		}
+	}
+	return "", ""
+}
+
+func getNewNsxtClient(node NsxClusterNode, d *schema.ResourceData, clients interface{}) (interface{}, error) {
+	// The API of joining a new node to cluster needs to be called with the new node's ip,
+	// to do that we initialize a new nsxtClient with new node's ip as its host
+	username := node.UserName
+	password := node.Password
+	host := node.IPAddress
+	if !strings.HasPrefix(host, "https://") {
+		host = fmt.Sprintf("https://%s", host)
+	}
+	c := clients.(nsxtClients)
+	commonConfig := c.CommonConfig
+	newClients := nsxtClients{
+		CommonConfig: commonConfig,
+	}
+	err := configureNewClient(&newClients, &c, host, username, password)
+	if err != nil {
+		return nil, err
+	}
+	return newClients, nil
+}
+
+func configureNewClient(newClient *nsxtClients, oldClient *nsxtClients, host string, username string, password string) error {
+	newClient.Host = host
+	securityCtx, err := getConfiguredSecurityContext(newClient, "", "", "", username, password)
+	if err != nil {
+		return fmt.Errorf("Failed to configure new client with host %s: %s", host, err)
+	}
+	newClient.PolicySecurityContext = securityCtx
+	newClient.PolicyHTTPClient = oldClient.PolicyHTTPClient
+	err = initNSXVersion(getPolicyConnector(*newClient))
+	if err != nil {
+		return fmt.Errorf("Failed to configure new client with host %s: %s", host, err)
+	}
+	return nil
+}
+
+func joinNodeToCluster(clusterID string, certSha256Thumbprint string, guestNode NsxClusterNode, masterNodeIP string, d *schema.ResourceData, m interface{}) error {
+	c, err := getNewNsxtClient(guestNode, d, m)
+	if err != nil {
+		return err
+	}
+	newNsxClients := c.(nsxtClients)
+	connector := getPolicyConnector(newNsxClients)
+	client := nsx.NewClusterClient(connector)
+	username, password := getHostCredential(m)
+	joinClusterParams := nsxModel.JoinClusterParameters{
+		CertificateSha256Thumbprint: &certSha256Thumbprint,
+		ClusterId:                   &clusterID,
+		IpAddress:                   &masterNodeIP,
+		Username:                    &username,
+		Password:                    &password,
+	}
+	_, err = client.Joincluster(joinClusterParams)
+	if err != nil {
+		return fmt.Errorf("Failed to join node to cluster: %s, node ip address: %s", err, guestNode.IPAddress)
+	}
+	return nil
+}
+
+func getHostCredential(m interface{}) (string, string) {
+	nsxtClient := m.(nsxtClients)
+	username := nsxtClient.NsxtClientConfig.UserName
```

Comment:
`nsxtClientConfig` is a struct used to initialize old sdk. We shouldn't rely on it, please use `CommonConfig` to store provider data intended for later use.